Mathlib/Tactic/HashCommandLinter.lean

```diff
@@ -0,0 +1,56 @@
+/-
+Copyright (c) 2024 Damiano Testa. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Damiano Testa
+-/
+
+import Std.Lean.Command
+import Std.Data.Array.Basic
+
+/-!
+#  `#`-command linter
+
+The `#`-command linter produces a warning when a command starting with `#` is used.
+-/
+
+namespace Mathlib.Linter
+
+/--
+Enables the '`#`-command' linter. This will warn on any command beginning with `#`.
+For example, `#eval, #guard, #guard_msgs` all trigger a message.
+-/
+register_option linter.hashCommand : Bool := {
+  defValue := true
```

Comment:
I don't think this should be enabled by default, but it should be enabled in CI (can do so in this PR). What do you think?

Reply:
I actually think that I prefer it to be disabled by default outside of CI: if I am using `#eval` in a file, I expect to want to see the output and not be bugged by the linter!  I'll look up how to make sure that CI uses a different default.

Reply:
Yeah I agree exactly, I think its enoguh to call lean in CI with `-Dlinter.hashCommand=true`

Reply:
I am failing to make this work locally:
```
$ lake build Mathlib.mwe
[303/304] Building Mathlib.mwe
error: > LEAN_PATH=./.lake/packages/std/.lake/build/lib:./.lake/packages/Qq/.lake/build/lib:./.lake/packages/aesop/.lake/build/lib:./.lake/packages/proofwidgets/.lake/build/lib:./.lake/packages/Cli/.lake/build/lib:./.lake/packages/importGraph/.lake/build/lib:./.lake/build/lib LD_LIBRARY_PATH=./.lake/build/lib /home/damiano/.elan/toolchains/leanprover--lean4---v4.6.0-rc1/bin/lean -Dlinter.hashCommand=true -Dpp.unicode.fun=true -Dpp.proofs.withType=false -DautoImplicit=false -DrelaxedAutoImplicit=false ./././Mathlib/mwe.lean -R ././. -o ./.lake/build/lib/Mathlib/mwe.olean -i ./.lake/build/lib/Mathlib/mwe.ilean -c ./.lake/build/ir/Mathlib/mwe.c
error: stderr:
invalid -D parameter, unknown configuration option 'linter.hashCommand'
error: external command `/home/damiano/.elan/toolchains/leanprover--lean4---v4.6.0-rc1/bin/lean` exited with code 1
```

Reply:
I modified this
```lean
  weakLeanArgs :=
    if get_config? CI |>.isSome then
      #["-DwarningAsError=true", "-Dlinter.hashCommand=true"]
    else
      #["-Dlinter.hashCommand=true"]
```

Reply:
Alex, it seems that it is not an option to make this linter quiet locally, but chatty on CI.  I propose to make the linter available with quiet default: hopefully in the future it will be possible to set different defaults globally and then the linter would reach its full potential!

Reply:
Hmm, it's a hack to work around it now, but: what if we made the `defValue := true`, yet *also* checked `warningAsError` in the linter (and left a very visible note, and even made the description ``"enables the `#`-command linter as long as `warningAsError` is `true`"``)? I.e. `getLinterHash (← getOptions) && warningAsError.get (← getOptions) && ...`. Then it run in CI (since that uses `warningAsError := true`) but not locally under normal circumstances.

I guess the question is: in practice, would someone want to turn on this linter locally? Or is it ok if it's only ever run in CI?

(Another option is to keep `defValue := false` but turn it on whenever `warningAsError := true`.)

Reply:
For me, the main use of this linter while working on a PR, would be when a `#`-command is silent.  If I am using a "noisy" `#`-command, I am probably curious about its output and adding the linter warning would be annoying.

If there were a way of making the linter noisy exactly when the `#`-command is not, then I would like it always on, I think.

Reply:
Having it on while editing when there are no messages is a good idea! I do wonder whether it should *also* be on everywhere during CI for good error messages (as mentioned elsewhere). I guess the logic would be `getLinterHash (<- getOptions) && ((<- get).messages.msgs.size == 0 || warningAsError.get (<- getOptions))` in that case. I won't insist on this, though! :)